environment/src/main/java/com/sequenceiq/environment/environment/service/PollerCollection.java

```diff
@@ -0,0 +1,216 @@
+        return () -> {
+            List<String> remaining = new ArrayList<>();
+            if (PollGroup.CANCELLED.equals(EnvironmentInMemoryStateStore.get(envId))) {
+                LOGGER.info("Datahub polling cancelled in inmemory store, id: " + envId);
+                return AttemptResults.breakFor("Datahub polling cancelled in inmemory store, id: " + envId);
+            }
+            List<AttemptResult<Void>> results = pollingCrn.stream()
+                    .map(crn -> fetchStartDistroXResult(remaining, crn))
+                    .collect(Collectors.toList());
+            return evaluateResult(new ArrayList<>(pollingCrn), remaining, results);
+        };
+    }
+
+    private AttemptResult<Void> fetchStartDistroXResult(List<String> remaining, String crn) {
+        StackV4Response stack = distroXV1Endpoint.getByCrn(crn, Collections.emptySet());
+        if (stackAndClusterAvailable(stack, stack.getCluster())) {
+            return AttemptResults.finishWith(null);
+        } else {
+            remaining.add(crn);
+            return checkStartStatus(stack);
+        }
+    }
+
+    public AttemptMaker<Void> stopDatalakePoller(List<String> pollingCrn, Long envId) {
+        return () -> {
+            if (PollGroup.CANCELLED.equals(EnvironmentInMemoryStateStore.get(envId))) {
+                LOGGER.info("Datalake polling cancelled in inmemory store, id: " + envId);
+                return AttemptResults.breakFor("Datalake polling cancelled in inmemory store, id: " + envId);
+            }
+            List<String> remaining = new ArrayList<>();
+            List<AttemptResult<Void>> results = pollingCrn.stream()
+                    .map(crn -> fetchStopDatalakeResult(remaining, crn))
+                    .collect(Collectors.toList());
+            return evaluateResult(new ArrayList<>(pollingCrn), remaining, results);
+        };
+    }
+
+    private AttemptResult<Void> fetchStopDatalakeResult(List<String> remaining, String crn) {
+        SdxClusterResponse sdx = sdxEndpoint.getByCrn(crn);
+        if (sdxStopped(sdx)) {
+            return AttemptResults.finishWith(null);
+        } else {
+            remaining.add(crn);
+            return checkStopStatus(sdx);
+        }
+    }
+
+    public AttemptMaker<Void> startDatalakePoller(List<String> pollingCrn, Long envId) {
+        return () -> {
+            if (PollGroup.CANCELLED.equals(EnvironmentInMemoryStateStore.get(envId))) {
+                LOGGER.info("Datalake polling cancelled in inmemory store, id: " + envId);
+                return AttemptResults.breakFor("Datalake polling cancelled in inmemory store, id: " + envId);
+            }
+            List<String> remaining = new ArrayList<>();
+            List<AttemptResult<Void>> results = pollingCrn.stream()
+                    .map(crn -> fetchStartDatalakeResult(crn, remaining))
+                    .collect(Collectors.toList());
+            return evaluateResult(new ArrayList<>(pollingCrn), remaining, results);
```

Comment:
Wrapping is repeated many times, and it's delegated to the callers of `evaluateResult`. It's just an internal behaviour but it would still be better if `evaluateResult` would be responsible for wrapping the things that needs to be wrapped for itself. Then you don't have to wrap them 4 times. It's error prone an requires attention unnecessarily.

Reply:
I have to revert all of ArrayList change because does not work